[WIP] refactored authorization event/event listeners

MichaelKubovic opened the pull request.

ajgarlag committed:
Add support for the "authorization_code" grant type

ajgarlag committed:
Add some integration tests

ajgarlag committed:
Add final integration tests

ajgarlag committed:
Use an event to resolve authorization request

An event listener shoud allow or deny the authorization request, or provide
an URI where the user will be redirected to view the authorization interface

ajgarlag committed:
Minor fixes

ajgarlag committed:
Remove unused dependencies

ajgarlag committed:
Use ScopeConverter to convert scopes

ajgarlag committed:
Improve naming and semantics

ajgarlag committed:
Move the authorization check into an EventListener

ajgarlag committed:
Remove authorization check

ajgarlag committed:
Fix README

ajgarlag committed:
Rename `decisionUri` to `resolutionUri`

This Uri can represent any URI, not only the decision (consent) uri.

ajgarlag committed:
Add return types to AuthorizationRequestResolveEvent

MichaelKubovic committed:
refactored authoriztion controller to independent event listeners

# Conflicts:
#	OAuth2Events.php

MichaelKubovic committed:
adds authorization strategy configuration

ajgarlag committed:
Use Security class instead of TokenStorageInterface

ajgarlag committed:
Use integers instead of booleans values for authorization resolution.

ajgarlag committed:
Test authorization code request with faked redirect uri

ajgarlag committed:
Rename AuthCode to AuthorizationCode

Classes under the `Trikoder\Bundle\OAuth2Bundle\League` remain named as
AuthCode to reflect the naming choosen by the league/oauth2-server project.

ajgarlag committed:
Method AbstractIntegrationTest::handleAuthorizeRequest returns Response instead of array

ajgarlag commented:
@MichaelKubovic Can you resolve conflict in order to review your PR?

Thanks!

ajgarlag committed:
Rename test methods

ajgarlag committed:
Update the AuthorizationRequestResolveEvent workflow

MichaelKubovic committed:
Merge remote-tracking branch 'ajgarlag/feature/authorization_code' into feature/authorization_code

# Conflicts:
#	Controller/AuthorizationController.php
#	Event/AuthorizationRequestResolveEvent.php
#	Resources/config/services.xml

MichaelKubovic commented:
Sorry for the delay. Done, please have a look. If https://github.com/trikoder/oauth2-bundle/pull/18 gets merged anytime soon, I could move the PR there. But for now I'd appreciate your feedback as it is.

ajgarlag reviewed (commented):
The main difference I see is that in my implementation the `AuthorizationRequest` instance is never mutated from the event listeners, but with your implementation, only two modification are allowed (setting the user and approving the request). I think this can be proposed to the upstream repository once https://github.com/trikoder/oauth2-bundle/pull/18 gets merged. Which I want to avoid is to expose the `AuthorizationRequest` instance from the event.

Another difference is that in my implementation, when an event listener decide about the `AuthorizationRequest` (allowing or denying it), the response is cleared so the user is redirected only if there is no decision.

I do not like the `authorization_strategy` related changes for different reasons:
  - This functionality can be implemented without introducing another level of indirection. Simply configure the appropriate event listeners.
  - To implement a new strategy, the bundle configuration should be modified, so it is difficult to extend it. If we only use event listeners, this problem is avoided.
  - If I choose the `user_consent` strategy, I'd like to set the `$consentApprovalRoute` from the bundle configuration, so the bundle config gets more complicated.

My thought was to left to the bundle user to decide how to implement the authorization process, but I think it can be convenient to provide an `AlwaysAllowAuthorizationRequestListener`, that could be easily enabled from a bundle configuration key, so we provide a minimal working implementation.

MichaelKubovic commented:
I should have stated the motivation behind this PR first:
Firstly, moving all `AuthorizationRequest` mutations to listeners has many benefits:
- it's makes the Authorization controller a mere dispatcher with the focus on the final state of the event
- having a well-defined sequence of listeners increases readability (what mutations happens when)
- it allows anybody to use CompilerPass to change only pieces they are interested in rather then overriding the whole controller
- plays nice with the plug&play listeners (I had the requirement to add the authentication listener myself)
I wanted to make a clear distinction between the internal design (listeners acting on the request) and a very focused extension points (how the authorization is done)

> The main difference I see is that in my implementation the `AuthorizationRequest` instance is never mutated from the event listeners, but with your implementation, only two modification are allowed (setting the user and approving the request). I think this can be proposed to the upstream repository once [trikoder#18](https://github.com/trikoder/oauth2-bundle/pull/18) gets merged. Which I want to avoid is to expose the `AuthorizationRequest` instance from the event.

Agree, encapsulation around `AuthorizationRequest` is kept in-tact.

> Another difference is that in my implementation, when an event listener decide about the `AuthorizationRequest` (allowing or denying it), the response is cleared so the user is redirected only if there is no decision.

Could you please point me to the lines where you clear the response? The method signature is intentionally strict `setResponse(Response $response)`, to disallow null response. Anybody willing to change the decision about whether the response should be set or not should do something about it in a listener that set the response in the first place. Listeners should honour the decision of the listeners that have already been executed on the request.

I would argue that allowing other developers to disapprove the request once it has been approved from the listener is not nice. You should override the listener that gave the approval, or use some of its extension point to augment the decision process. Hence the `approveAuthorization()`.

At the end of the day, we could do all of this in the AuthorizationController, but we break it down to give each class it's single responsibility and allow for greater extensibility and all that....

> 
> I do not like the `authorization_strategy` related changes for different reasons:
> 
> * This functionality can be implemented without introducing another level of indirection. Simply configure the appropriate event listeners.
That level of indirection is the extension point for a well-defined part of the OAuth spec. The built-in delegating listener makes it clear that there is a step that handles authorization decision in the OAuth2 process, but does not care about the way it's handled. It just tells that there is this authorization task that must be fulfilled by `AuthorizationDecisionStrategy`. It's possible to make it simpler by changing the strategies into separate listeners and switch the listeners based on configuration but then you could not force the `AuthorizationDecisionStrategy` interface. 

> * To implement a new strategy, the bundle configuration should be modified, so it is difficult to extend it. If we only use event listeners, this problem is avoided.
That's just my oversight, can be fixed easily.

> * If I choose the `user_consent` strategy, I'd like to set the `$consentApprovalRoute` from the bundle configuration, so the bundle config gets more complicated.
True that.

> My thought was to left to the bundle user to decide how to implement the authorization process, but I think it can be convenient to provide an `AlwaysAllowAuthorizationRequestListener`, that could be easily enabled from a bundle configuration key, so we provide a minimal working implementation.

I want to allow anybody to design their own Authorization controller/view - but keep the rest on us. The issue with `AlwaysAllowAuthorizationRequestListener` is that you need to remove it from the container so that it doesn't run. Whereas in my implementation, you switch the strategy to your custom class that implements the interface and you are good to go. You don't have to understand the bundle to the level that listeners are the mechanism that augment the authorization request.


I'll push changes that hopefully fix most of your concerns, but I'd like to keep the strategy thing. Open to renaming it though.

MichaelKubovic committed:
configure authentication listener properly

# Conflicts:
#	DependencyInjection/Configuration.php
#	DependencyInjection/TrikoderOAuth2Extension.php

MichaelKubovic committed:
use PSR response interface, reorder methods

MichaelKubovic committed:
configure consent strategy correctly

MichaelKubovic commented on Event/Listener/AuthorizationRequestAuthenticationListener.php:
```diff
@@ -0,0 +1,88 @@
+        SessionInterface $session,
+        RequestStack $requestStack,
+        UrlGeneratorInterface $urlGenerator,
+        string $loginRoute = 'app_login'
```
added configuration option

ajgarlag replied:
Where is this configuration option?

MichaelKubovic commented on Event/Listener/AuthorizationRequestAuthenticationListener.php:
```diff
@@ -0,0 +1,88 @@
+        }
+
+        if (!$this->authorizationChecker->isGranted('IS_AUTHENTICATED_REMEMBERED')) {
+            $this->saveTargetPath($this->session, 'main', $request->getUri());
```
done, current firewall is retrieved from `FirewallMap`

MichaelKubovic commented on Event/Listener/AuthorizationRequestAuthenticationListener.php:
```diff
@@ -0,0 +1,88 @@
+            throw new \RuntimeException('Authentication listener depends on the request context');
+        }
+
+        if (!$this->authorizationChecker->isGranted('IS_AUTHENTICATED_REMEMBERED')) {
```
It does not really make sense to require IS_AUTHENTICATED_FULLY during OAuth authorization flow. That would require the user to enter his credentials any time he comes to authorization endpoint (this goes against the single sign-on I'm trying to build). However, if there will be a real requirement driven by the use-case, we can still make it into a configuration option later.

ajgarlag replied:
It makes sense for me if I store the client authorization to a persistent storage, so the user only have to consent once for each client (if scopes do not change). In that case, I want to be sure that the user is fully authenticated. I think GitHub authorization workflow works in a similar way.

MichaelKubovic commented on Event/AuthorizationRequestResolveEvent.php:
```diff
@@ -7,64 +7,32 @@
     private $authorizationRequest;
 
     /**
-     * @var ?ResponseInterface
```
sure, my bad, fixed

MichaelKubovic commented on Event/AuthorizationRequestResolveEvent.php:
```diff
@@ -7,64 +7,32 @@
-        $this->response = null;
-    }
-
-    public function hasResponse(): bool
```
my bad, fixed

MichaelKubovic committed:
reordered methods

MichaelKubovic commented on Model/AuthorizationDecision/UserConsentDecisionStrategy.php:
```diff
@@ -0,0 +1,169 @@
+        UriSigner $uriSigner,
+        RequestStack $requestStack,
+        UrlGeneratorInterface $urlGenerator,
+        string $consentApprovalRoute = 'oauth2_consent'
```
configuration option added

MichaelKubovic commented on DependencyInjection/Configuration.php:
```diff
@@ -58,6 +58,11 @@ private function createAuthorizationServerNode(): NodeDefinition
                 ->end()
+                ->enumNode('authorization_strategy')
+                    ->info("What strategy should be used to authorize user.\nAvailable options are: user_consent - let user authorize the client request, always_allow - silently authorizes all client requests")
+                    ->values(['user_consent', 'always_allow'])
```
I've changed it so that `authorization_strategy` accepts service id.

ajgarlag commented:
IMO there should be three ways to resolve an authorization request:
  1. Approve it
  2. Deny it (not allowed by your implementation)
  3. Redirect the user to a consent endpoint (setting a response)

With your implementation, if a listener set a response to the event, the user will receive that response; no matter if another listener approves the authorization later. And if a listener approves the authorization, another listener cannot deny it.

I prefer my `AuthorizationRequestResolveEvent::resolveAuthorization(bool $authorizationResolution)` which can allow or deny the authorization request, and clear the response if any; over your `AuthorizationRequestResolveEvent::approveAuthorization`. I think it's more flexible. If you stink think that the decision made from a listener should not be overriden from another one, you can call `Event::stopPropagation` from that listener.

MichaelKubovic commented:
> If you stink think that the decision made from a listener should not be overriden from another one, you can call `Event::stopPropagation` from that listener.

Would that skip all the other built-in listeners that you actually want to run?

ajgarlag commented on Event/AuthorizationRequestResolveEvent.php:
```diff
@@ -9,49 +9,36 @@
-    public function resolveAuthorization(bool $authorizationResolution): void
-    {
-        $this->authorizationResolution = $authorizationResolution;
-        $this->response = null;
```
Here, if a listener approves or denies the authorization request, the response is unset.

ajgarlag commented on Event/AuthorizationRequestResolveEvent.php:
```diff
@@ -9,49 +9,36 @@
     public function hasResponse(): bool
     {
-        return $this->response instanceof ResponseInterface;
+        return $this->response !== null;
```
Why this change?

ajgarlag commented on Event/AuthorizationRequestResolveEvent.php:
```diff
@@ -9,49 +9,36 @@
     }
 
-    public function getResponse(): ResponseInterface
+    public function getResponse(): ?ResponseInterface
```
This method wont return null.

ajgarlag commented:
> > If you stink think that the decision made from a listener should not be overriden from another one, you can call `Event::stopPropagation` from that listener.
> 
> Would that skip all the other built-in listeners that you actually want to run?

Yes, so I prefer to allow any listener to override a previous decision. But with your implementation, if a listener has set a response, another one cannot approve the request later, so I think in that case, you should make it explicit calling `Event::stopPropagation`.

With my `AuthorizationRequestResolveEvent::resolveAuthorization(bool $authorizationResolution)` we avoid that problem. But maybe it's not clear that the authorization request can be resolved calling to `resolveAuthorization` or `setResponse`.

What do you think about adding a `AuthorizationRequestResolveEvent::resolve(bool|ResponseInterface $resolution)` method and removing the `AuthorizationRequestResolveEvent::setResponse` method. With that implementation, the last call to the `resolve` method would win the authorization resolution.

MichaelKubovic commented:
Right, I prefer not to run listener that has no impact (because its outcome is supersede by the succeeding listener). It seems more natural to change the way the authorization is resolved, to fix the decision process that produced the original resolution in the first place. When you just add another listener that changes previous decision - what is the point in running the original listener at all?

I see some analogy with Gaurd authenticators. When some of them set the response, they are done.
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php#L91

MichaelKubovic commented:
A use-case for the listener that overrides previously generated response or authorization resolution  would help me understand your motivation.
